pkg/asset/installconfig/gcp/validation.go

```diff
@@ -141,66 +141,144 @@ func validateZoneProjects(client API, ic *types.InstallConfig, fieldPath *field.
 	return allErrs
 }
 
-// validateManagedZones will validate the public and private managed zones if they exist.
-func validateManagedZones(client API, ic *types.InstallConfig, fieldPath *field.Path) field.ErrorList {
-	commonZoneLookup := func(dnsZone, zoneProject, defaultProject string) error {
-		project := zoneProject
-		if project == "" {
-			project = defaultProject
-		}
+// internalFindProject finds the correct project to use during installation. If the project id is
+// provided in the zone use the project id, otherwise use the default project.
+func internalFindProject(zone *gcp.DNSZone, defaultProject string) string {
+	if zone != nil && zone.ProjectID != "" {
+		return zone.ProjectID
+	}
+	return defaultProject
+}
 
-		if _, err := client.GetDNSZoneByName(context.TODO(), project, dnsZone); err != nil {
-			return err
-		}
-		return nil
+// internalFindDNSZone finds a zone in a project. If a project is provided in the zone, the project
+// is used otherwise the default project is used.
+func internalFindDNSZone(client API, zone *gcp.DNSZone, defaultProject string) (*dns.ManagedZone, string, error) {
+	if zone == nil {
+		return nil, defaultProject, fmt.Errorf("zone is not set")
 	}
 
+	project := internalFindProject(zone, defaultProject)
+	returnedZone, err := client.GetDNSZoneByName(context.TODO(), project, zone.ID)
+	if err != nil {
+		return nil, project, err
+	}
+	return returnedZone, project, nil
+}
+
+// validateManagedZones will validate the public and private managed zones if they exist.
+func validateManagedZones(client API, ic *types.InstallConfig, fieldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 
 	if ic.GCP.PublicDNSZone != nil && ic.GCP.PublicDNSZone.ID != "" {
-		if err := commonZoneLookup(ic.GCP.PublicDNSZone.ID, ic.GCP.PublicDNSZone.ProjectID, ic.GCP.ProjectID); err != nil {
-			allErrs = append(allErrs, field.Invalid(fieldPath.Child("PublicDNSZone").Child("ID"), ic.GCP.PublicDNSZone.ID, "invalid public managed zone"))
+		//_, _, err := internalFindDNSZone(client, ic.Platform.GCP.PublicDNSZone, ic.GCP.ProjectID)
+		//if err != nil {
+		//	return allErrs
+		//}
+		returnedZone, project, err := internalFindDNSZone(client, ic.Platform.GCP.PublicDNSZone, ic.GCP.ProjectID)
```

Comment:
Passing the project in and then getting the project (or a different project) out I think makes this difficult to read. I think it would be better to either call `internalFindProject` separately or implement my previous suggestion of using a default value for the project.